blog/content/post/2021-10-28-rockchip-secure-boot.md

```diff
@@ -0,0 +1,364 @@
+
+```shell
+$ ../tools/linux/rk_sign_tool/rk_sign_tool sl --key ../keys/privateKey.pem --pubkey ../keys/publicKey.pem --loader rk3288_loader_v1.09.258.bin
+start to sign rk3288_loader_v1.09.258.bin
+path = /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader
+unpacking loader start...
+loader=rk3288_loader_v1.09.258.bin
+output=/hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/
+unpacking UsbDataVRL...
+unpacking rk3288_ddr_400MHz_v1...
+unpacking rk3288_usbplug_v2...
+unpacking FlashData...
+unpacking FlashBoot...
+unpack loader ok.
+packing loader start...
+writing entry...
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/UsbDataVRL.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/rk3288_ddr_400MHz_v1.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/rk3288_usbplug_v2.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/FlashData.bin
+writing /hdd/rk-secure-boot/tools/linux/rk_sign_tool/temp/loader/FlashBoot.bin
+writing crc...
+pack loader ok.(rk3288_loader_v1.09.258.bin)(0.01)
+sign loader ok.
+```
+
+But when I tried loading it into eFUSE Tool I've got an error.
+
+![Firmware loading failed](/img/rk_efuse_tool_error.png)
+
+That was an extremely helpful message. To find out what's wrong, I've done a
+quick analysis of unsigned and signed binary and discovered that the signed
+version has a different header - this is how the unsigned version looks like.
+
+```
+00000000  42 4f 4f 54 66 00 3a 02  00 00 00 00 00 01 e5 07  |BOOTf.:.........|
+00000010  0a 1c 10 15 00 41 30 32  33 01 66 00 00 00 39 01  |.....A023.f...9.|
+```
+
+And this is the signed version.
+
+```
+00000000  4c 44 52 20 66 00 3a 02  00 00 00 00 00 01 e5 07  |LDR f.:.........|
+00000010  0a 1c 10 15 24 41 30 32  33 02 66 00 00 00 39 01  |....$A023.f...9.|
+```
+
+Secure Boot Tool has support for signing loader binary, but it is disabled when
+program starts. I had to dig through RockChip documentation to find out how to
+enable it. Hint: press CTRL+R+K.
+
+![Secure Boot Tool](/img/rk_secure_boot_tool.png)
+
+Signing with Secure Boot Tool produced binary with a correct header. I can load
+it into eFUSE Tool also. Note that even though eFUSE Tool cannot load binaries
+built by `rk_sign_tool`, `rkdeveloptool` works with them just fine.
+
+## Burning eFUSE
+
+This was the most dangerous part as I had only try. At that time, I didn't know
+how eFUSE Tool operates (as there is no documentation) - it burns only public
+key extracted from binary, or burns key and flashes binary onto eMMC? What will
+happen if I get stuck with loader flashed onto eMMC but no U-Boot to load? Does
+this tool work at all? Will it burn the correct key or some random trash? If
+fusing succeeds, will be MaskROM mode still available?
+
+RockChip always boots from SPI first, then eMMC, then SD card. There is no way
+to alter boot order on these SoCs. The only way to recover from freezing boot is
```

Comment:
these? which?

Reply:
I changed text a bit to avoid using this word.